push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/PushyApnsSender.java

```diff
@@ -0,0 +1,223 @@
+import com.relayrides.pushy.apns.util.ApnsPayloadBuilder;
+import io.netty.util.concurrent.Future;
+import org.jboss.aerogear.unifiedpush.api.Variant;
+import org.jboss.aerogear.unifiedpush.api.VariantType;
+import org.jboss.aerogear.unifiedpush.api.iOSVariant;
+import org.jboss.aerogear.unifiedpush.message.InternalUnifiedPushMessage;
+import org.jboss.aerogear.unifiedpush.message.Message;
+import org.jboss.aerogear.unifiedpush.message.UnifiedPushMessage;
+import org.jboss.aerogear.unifiedpush.message.apns.APNs;
+import org.jboss.aerogear.unifiedpush.message.sender.apns.AeroGearApnsPushNotification;
+import org.jboss.aerogear.unifiedpush.service.ClientInstallationService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.io.ByteArrayInputStream;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.ConcurrentSkipListSet;
+import java.util.concurrent.ExecutionException;
+
+import static org.jboss.aerogear.unifiedpush.message.util.ConfigurationUtils.tryGetIntegerProperty;
+import static org.jboss.aerogear.unifiedpush.message.util.ConfigurationUtils.tryGetProperty;
+
+@SenderType(VariantType.IOS)
+public class PushyApnsSender implements PushNotificationSender {
+
+    private final Logger logger = LoggerFactory.getLogger(PushyApnsSender.class);
+
+    public static final String CUSTOM_AEROGEAR_APNS_PUSH_HOST = "custom.aerogear.apns.push.host";
+    public static final String CUSTOM_AEROGEAR_APNS_PUSH_PORT = "custom.aerogear.apns.push.port";
+    private static final String customAerogearApnsPushHost = tryGetProperty(CUSTOM_AEROGEAR_APNS_PUSH_HOST);
+    private static final Integer customAerogearApnsPushPort = tryGetIntegerProperty(CUSTOM_AEROGEAR_APNS_PUSH_PORT);
+
+    @Inject
+    private ClientInstallationService clientInstallationService;
+    private final ConcurrentSkipListSet<String> invalidTokens = new ConcurrentSkipListSet();
+
+
+    @Override
+    public void sendPushMessage(final Variant variant, final Collection<String> tokens, final UnifiedPushMessage pushMessage, final String pushMessageInformationId, final NotificationSenderCallback senderCallback) {
+        // no need to send empty list
+        if (tokens.isEmpty()) {
+            return;
+        }
+
+        final iOSVariant iOSVariant = (iOSVariant) variant;
+
+        final String payload;
+        {
+            try {
+                payload = createPushPayload(pushMessage.getMessage(), pushMessageInformationId);
+            } catch (IllegalArgumentException iae) {
+                logger.info(iae.getMessage());
+                senderCallback.onError("Nothing sent to APNs since the payload is too large");
+                return;
+            }
+        }
+
+        final ApnsClient apnsClient = buildApnsClient(iOSVariant);
+
+        // connect:
+        final Future<Void> connectFuture = connectToDestinations(iOSVariant, apnsClient);
+        try {
+            connectFuture.await();
```

Comment:
So, I'd need to move the "send" inside of that callback, because otherwise there is different exception (e.g. ClientNotConnected).

let me update